### PR TITLE
[MNGSITE-491] Update commit policy related document

### PR DESCRIPTION
**Pull request by cstamas**
Last update to ASF Maven commit policy happened while project sources were hosted in Subversion repository... Update relevant bits w/ respect to changes happened since then.

---

https://issues.apache.org/jira/browse/MNGSITE-491

### CODE REVIEW

**Review by khmarbaise — APPROVED** [review]

**Comment by cstamas** [comment]
@khmarbaise eh, you were too fast, I wanted to add one remark:

@khmarbaise (w/ your PMC chair hat on): am unsure was this ever voted or decided upon, but I think it is **undoubted** that WHAT we do since few years and WHAT site states is out of sync. So, I'd leave it up to you, if needed, to act on it (ie. start a vote if needed).

File: content/markdown/project-roles.md [hunk 1]
```diff
@@ -134,11 +134,25 @@ These are those people who have been given write access to the
...
+Committers are responsible for ensuring that every file they
+commit is covered by a valid CLA.
 
-Committers are responsible for ensuring that every file they 
-commit is covered by a valid CLA. 
+The Apache Maven project follows both ASF commit policies as described in
+[ASF Development Process](https://www.apache.org/foundation/how-it-works/legal.html)
+document.
+
+In general, "review then commit" (RTC) is required for complex changes or against "stable"
```
Comment (slawekjaranowski) [inline content/markdown/project-roles.md:144]: simple changes which change behavior - should be reviewed also

File: content/markdown/project-roles.md [hunk 2]
```diff
@@ -134,11 +134,25 @@ These are those people who have been given write access to the
...
-Committers are responsible for ensuring that every file they 
-commit is covered by a valid CLA. 
+The Apache Maven project follows both ASF commit policies as described in
+[ASF Development Process](https://www.apache.org/foundation/how-it-works/legal.html)
+document.
+
+In general, "review then commit" (RTC) is required for complex changes or against "stable" 
+branches, while "commit then review" (CTR) is allowed for the rest of changes. 
+Always inform yourself (ultimately by asking on developers mailing list) which branches 
+are considered as "stable" and which as not stable ones.
```
Comment (slawekjaranowski) [inline content/markdown/project-roles.md:147]: most of components have only one master branch, like plugins shares components

In case when we have many branches we can put such info in project README in branch

**Review by michael-o — APPROVED** [review]
Oh, yes!

File: content/apt/developers/conventions/git.apt [hunk 3]
```diff
@@ -64,8 +64,12 @@ Maven Git Convention
 ** For committers
 
   Committers may, of course, commit directly to the ASF repositories.
-  For complex changes, you may find it valuable to make a pull request
-  at github to make it easier to collaborate with others.
+  For complex changes or changes against stable branches it is required
```
Comment (olamy) [inline content/apt/developers/conventions/git.apt:67]: maybe need to define what is `stable` branch? do you mean `master` for every plugin?  
or refer to the other page to define the meaning to avoid different content in different places

**Review by olamy — CHANGES_REQUESTED** [review]

File: content/apt/developers/conventions/git.apt (hunk 3 above)
Comment (cstamas) [reply content/apt/developers/conventions/git.apt:67]: See other changed file, the git convention is below that page.

**Review by slawekjaranowski — APPROVED** [review]
Generally OK. 

We can improve eg. "Developers Conventions" and show in more details - what is simple change, how work with reviews etc ...

File: content/apt/developers/conventions/git.apt (hunk 3 above)
Comment (olamy) [reply content/apt/developers/conventions/git.apt:67]: this will be in another page when the site will deployed.
that's why I said `refer to the other page to define the meaning to avoid different content in different places`

**Comment by cstamas** [comment]
My take on "stable branch": a branch that was used for, or is decided that will be used for release". Hence, in case of maven repo, this means: master (future Maven4), maven-3.9.x (we plan 3.9,x release) and maven-3.8.x (was used to release 3.8.x so far).

Still, this PR defines stable branch as "ask and we will tell you" :smile:

File: content/apt/developers/conventions/git.apt (hunk 3 above)
Comment (cstamas) [reply content/apt/developers/conventions/git.apt:67]: Added link to other page.

**Comment by cstamas** [comment]
@olamy pls review latest changes and either remove your "change requested" status or comment what more needed.

File: content/apt/developers/conventions/git.apt (hunk 3 above)
Comment (michael-o) [reply content/apt/developers/conventions/git.apt:67]: Maybe stable is the wrong word, but rather 'maintained branches'?

File: content/apt/developers/conventions/git.apt [hunk 4]
```diff
@@ -64,8 +64,14 @@ Maven Git Convention
...
+  to make a pull request at github to make it easier to collaborate with
```
Comment (michael-o) [inline content/apt/developers/conventions/git.apt:68]: I patch with the JIRA ticket should also be fine. We shouldn't people force to use GitHub, if they don't want/can't.

File: content/markdown/project-roles.md (hunk 1 above)
Comment (michael-o) [reply content/markdown/project-roles.md:144]: Correct. Even a dependency upgrade *can* change the behavior. Non-functional changes should be safe to push.

**Review by michael-o — CHANGES_REQUESTED** [review]

**Comment by hgschmie** [comment]
I don't think that a pull request for a document is the right place to discuss project wide policy changes. This is a conversation for the PMC list (and the PMC might need to vote on it).

Policy in any Apache project is set by the PMC. Not through pull requests. I strongly recommend to respect the Apache conventions. This feels a lot like trying to circumvent the "inconvenient" process of finding consensus on the PMC. I am pretty sure that most PMC members are not even aware of this PR and at least I (if I were on the PMC) would be upset that this conversation happens outside.

@khmarbaise @aheritier @andham @bimargulies @brianf @cstamas @dennisl @dkulp @hboutemy @ifedorenko @jvanzyl @krosenvold @olamy @michael-o @rfscholte @rgoers @slawekjaranowski @stephenc @struberg @Tibor17 

I did not find / know github committer ids for Barrie Treloar, Emmanuel Venisse, Guillaume Boue, Milos Kleint, Vincent Sieveton and Wayne Fay. All of those are PMC members (according to https://maven.apache.org/team.html).

**Review by hgschmie — COMMENTED** [review]
Unless you feel that I should join the PMC, I will not comment on project policy.

**Comment by stephenc** [comment]
The PMC gets to decide the policy, but in my opinion... and I strongly suspect the rest of the PMC... everyone should be entitled to make suggestions. The PMC is really here to help the community... The PMC is the servant of each project's community... obviously we also have to operate within the bounds of the foundation... but IMO a PR is exactly the place to discuss this kind of thing... even if in the end the PMC's votes are the binding ones for policy

**Comment by olamy** [comment]
regarding this change some vote might be needed or some discussion on a mailing list....
not sure if it's mandatory per ASF rule but might need to be checked.
because you can definitely not see a github notification in the middle of all noise generated by GH.

btw if this makes mandatory to make a PR what about the requirement to merge a PR?
if moving to RTC the requirements to be able to merge a PR must defined.
number of people approving a PR, how many time before being able to merge, etc...
because frankly if making a PR is mandatory but then it's merged in 10 minutes with only one approval. 
I do not see the point except useless BS (pardon my french) procedures... 
but yeah on the other side this will considerably slow down development time... :(

**Comment by cstamas** [comment]
Folks, I thought I was precise enough with this PR, but seems I was not, so let me reiterate:
* this PR is NOT changing project policy, what it tries to do instead is to **align doco to what we've been doing for last 10 years** (basically since migrated to git reposes and Github got involved with ASF Maven reposes).
* Hence, from my POV this merely "updates" the (stale, last edited in SVN-times 13 years ago) documentation to what happens today.
* is not changing project stance regarding commit policy. Everyone who was on project for last 10 years (since migration to git and Github) can assert this IMHO. But clearly it's possible I am wrong, and Maven project processing PRs was just something "optional"?

In short: Maven project got way too big to let people just do CTR at their own will, especially today, when we keep 3 "stable" branches (3.8.x for eventual bugfixes, the upcoming 3.9.x that we are alrwady are delayed with, and upcoming 4.0.x). Not to mention all the hundreds of other reposes (as these above are "just" maven itself), and I believe we expect that the branches are in state to be released whenever needed...

**Comment by michael-o** [comment]
@cstamas after the Maven 3.4.0 debacle we switched to RTC for good reasons @stephenc remembers.

**Comment by stephenc** [comment]
> @cstamas after the Maven 3.4.0 debacle we switched to RTC for good reasons @stephenc remembers.

I remember 3.4.0 but I don't remember us switching to RTC.

What happened is that at the point of the vote, the PMC and others raised concerns additional to the normal stuff that we have to check at release time (ie are all commits covered by ICLAs, does the code in the source bundle represent what is being released, etc. see the requirements the ASF pushes down to the PMC for full details)

Because that release would break backwards compatibility, and because there had been a lot of stuff merged, the decision was made to revert back to the previous release and cherry pick forward.

But that was not an RTC policy, that was a pragmatic approach to the nature of the changes... and it had the negative effect of pushing an active committer away... even if ultimately the actions we took were for the benefit of the wider Maven community (which is what the PMC is responsible for nurturing)

**Comment by michael-o** [comment]
Our agreement (informal) was that changes to Corw go through PRs.

**Comment by olamy** [comment]
> Our agreement (informal) was that changes to Corw go through PRs.

yup but PRs mandatory != RTC vs CTR.

The technical fact of make it mandatory to provide a PR has nothing to do with the CTR vs RTC policy.
What is the difference between pushing directly and make a PR then merge it 5 minutes later without real review? (Just being able to say "hey I made a a PR"...)
Again I have nothing against adding such procedure (e.g PR mandatory) and it's probably good (btw majority of dev already do it).
But if changing CTR to RTC this has to be clearly defined because this can be very ambiguous and easy to play with and could be even worst... look at how it's already ambiguous today.

Another technical point is to leave a reference to the PR in the commit to be able to track the origin of the commit (e.g ensure this is coming from a PR)
Look at those changes in core master:
- https://github.com/apache/maven/commit/5502c0c8dba03c7241bd9dd1bb5ba023ab6f2670
- https://github.com/apache/maven/commit/750e1e3e71194644acda437cefbd73c8bd408f5d

how do you know those commits comes from a PR? the commit doesn't even have the same hash. 
Maybe I missed some git magic but Is there a way to be sure the commit is the same as the one tested in the PR?

**Comment by michael-o** [comment]
> > Our agreement (informal) was that changes to Corw go through PRs.
> 
> yup but PRs mandatory != RTC vs CTR.
> 
> The technical fact of make it mandatory to provide a PR has nothing to do with the CTR vs RTC policy. What is the difference between pushing directly and make a PR then merge it 5 minutes later without real review? (Just being able to say "hey I made a a PR"...) Again I have nothing against adding such procedure (e.g PR mandatory) and it's probably good (btw majority of dev already do it). But if changing CTR to RTC this has to be clearly defined because this can be very ambiguous and easy to play with and could be even worst... look at how it's already ambiguous today.
> 
> Another technical point is to leave a reference to the PR in the commit to be able to track the origin of the commit (e.g ensure this is coming from a PR) Look at those changes in core master:
> 
> * [apache/maven@5502c0c](https://github.com/apache/maven/commit/5502c0c8dba03c7241bd9dd1bb5ba023ab6f2670)
> * [apache/maven@750e1e3](https://github.com/apache/maven/commit/750e1e3e71194644acda437cefbd73c8bd408f5d)
> 
> how do you know those commits comes from a PR? the commit doesn't even have the same hash. Maybe I missed some git magic but Is there a way to be sure the commit is the same as the one tested in the PR?

You can assure it two ways: push to the branch before merge and then add the closes line.

**Comment by olamy** [comment]
> > > Our agreement (informal) was that changes to Corw go through PRs.
> > 
> > 
> > yup but PRs mandatory != RTC vs CTR.
> > The technical fact of make it mandatory to provide a PR has nothing to do with the CTR vs RTC policy. What is the difference between pushing directly and make a PR then merge it 5 minutes later without real review? (Just being able to say "hey I made a a PR"...) Again I have nothing against adding such procedure (e.g PR mandatory) and it's probably good (btw majority of dev already do it). But if changing CTR to RTC this has to be clearly defined because this can be very ambiguous and easy to play with and could be even worst... look at how it's already ambiguous today.
> > Another technical point is to leave a reference to the PR in the commit to be able to track the origin of the commit (e.g ensure this is coming from a PR) Look at those changes in core master:
> > 
> > * [apache/maven@5502c0c](https://github.com/apache/maven/commit/5502c0c8dba03c7241bd9dd1bb5ba023ab6f2670)
> > * [apache/maven@750e1e3](https://github.com/apache/maven/commit/750e1e3e71194644acda437cefbd73c8bd408f5d)
> > 
> > how do you know those commits comes from a PR? the commit doesn't even have the same hash. Maybe I missed some git magic but Is there a way to be sure the commit is the same as the one tested in the PR?
> 
> You can assure it two ways: push to the branch before merge and then add the closes line.

well I'm not sure if we need something too complicated as we can even "squash and merge" PRs. And we probably don't want to waste too much time with that.
Perso I like the default behaviour of GH such:
- https://github.com/apache/maven/commit/90c084dcf495e73c679395bc4b050500cbccf8cf
- https://github.com/apache/maven/commit/ce4579108d653be2ab7eab43be7d5951151dae5b
- 
so it's easy to click on a link or having a PR id to be sure this has been through a PR

**Comment by olamy** [comment]
thinking loud but what about something:

All changes must done via PRs.
PRs can be merged only if:
- CI build pass (some flakyness can be admitted due to overloading of some CI)

AND

- PR can be merged after 24h period review and with at least 1 approval
OR
- PR can be merged after 72h period review if nobody has reviewed it


Merged commit/squash/rebase must contain a reference to the PR

As usual any merged PR can be vetoed.
It's a sort of mix between CTR and RTC.
do we apply this to core and rest or only core and something different to the rest?

**Comment by michael-o** [comment]
> thinking loud but what about something:
> 
> All changes must done via PRs. PRs can be merged only if:
> 
>     * CI build pass (some flakyness can be admitted due to overloading of some CI)
> 
> 
> AND
> 
>     * PR can be merged after 24h period review and with at least 1 approval
>       OR

This I like.

>     * PR can be merged after 72h period review if nobody has reviewed it

This is problematic because three days is not much, especially when there is weekend, public holidays, etc. Consider also that Pr can be complex which means they require some significant amount of time to understand and evaluate. For me personally, three days is nothing.

> Merged commit/squash/rebase must contain a reference to the PR

Very good, this I have been personally doing for years. `git commit --amend`.

> As usual any merged PR can be vetoed. It's a sort of mix between CTR and RTC. do we apply this to core and rest or only core and something different to the rest?

All because non-core components are not less important. They can also cause subtile errors.

**Comment by stephenc** [comment]
> This is problematic because three days is not much, especially when there is weekend, public holidays, etc. Consider also that Pr can be complex which means they require some significant amount of time to understand and evaluate. For me personally, three days is nothing.

72h is the standard voting period of the ASF. I see no reason to have any vote by lazy consensus (which is effectively what this is) require a longer waiting period.

You still require a committer to actually pick up the PR if you don't have the commit bit.

And if you do have the commit bit, that means we trust you.

I like

**Comment by stephenc** [comment]
> CI build pass (some flakyness can be admitted due to overloading of some CI)

There are also cases where you need to make some changes that risk breaking the CI job itself such that a passing CI is impossible... think changes that affect the CI process itself... a change to the build process on the master branch may only be fully testable when it has been merged... and some changes invoke the "trusted branch" effect such that they cannot be verified on a PR...

Modulo those exceptions I think a passing CI is a requirement... but I would word it as:

* A passing CI build... or a valid documented reason why the CI build cannot be made to pass and in that case you no longer get to merge without at least one approval

**Comment by slawekjaranowski** [comment]
> This is problematic because three days is not much, especially when there is weekend, public holidays, etc. Consider also that Pr can be complex which means they require some significant amount of time to understand and evaluate. For me personally, three days is nothing.

You can always put a comment that you want to do review and need more time, add yourself as reviewer ... of course if you not absent

But I agree that standard 72H should be enough for first contact.

---

> Merged commit/squash/rebase must contain a reference to the PR

If we will do PR on GitHub, reference to PR are saved in commit, can be accessible by GUI or api

```
gh api /repos/apache/maven/commits/750e1e3/pulls --jq '.[].html_url'
https://github.com/apache/maven/pull/792
```

**Comment by olamy** [comment]
> > This is problematic because three days is not much, especially when there is weekend, public holidays, etc. Consider also that Pr can be complex which means they require some significant amount of time to understand and evaluate. For me personally, three days is nothing.
> 
> You can always put a comment that you want to do review and need more time, add yourself as reviewer ... of course if you not absent
> 
> But I agree that standard 72H should be enough for first contact.
> 
> > Merged commit/squash/rebase must contain a reference to the PR
> 
> If we will do PR on GitHub, reference to PR are saved in commit, can be accessible by GUI or api
> 
> ```
> gh api /repos/apache/maven/commits/750e1e3/pulls --jq '.[].html_url'
> https://github.com/apache/maven/pull/792
> ```

Ok. but not sure we really need some complicated script ;)
if some wants why not. 
but perso I merge using GH UI or `gh` cli tool. It's simple/just doing the job/ezy/convenient/etc... :)

**Comment by hgschmie** [comment]
I feel that forcing apache committers to use non-apache (gh) infrastructure by building a process that *requires* using github might be problematic. If I were on the PMC, I would probably ask the board about this.

**Comment by hgschmie** [comment]
> > This is problematic because three days is not much, especially when there is weekend, public holidays, etc. Consider also that Pr can be complex which means they require some significant amount of time to understand and evaluate. For me personally, three days is nothing.
> 
> You can always put a comment that you want to do review and need more time, add yourself as reviewer ... of course if you not absent

In that case, you can review (and push back) after the commit. I feel that 72h are a courtesy by every committer to allow review. But I reality, a committer has the *right* to commit to the code base without seeking additional approvals or permission or require review. That is the Apache way. Asking the committers very nicely (or write it down) to give a 72h hour review period before merging is perfectly acceptable. Requiring non-Apache tools is not. 

There is nothing wrong with committing directly to gitbox. https://infra.apache.org/apache-github.html claims that it is a dual-master sync setup.

**Comment by olamy** [comment]
> > > This is problematic because three days is not much, especially when there is weekend, public holidays, etc. Consider also that Pr can be complex which means they require some significant amount of time to understand and evaluate. For me personally, three days is nothing.
> > 
> > 
> > You can always put a comment that you want to do review and need more time, add yourself as reviewer ... of course if you not absent
> 
> In that case, you can review (and push back) after the commit. I feel that 72h are a courtesy by every committer to allow review. But in reality, a committer has the _right_ to commit to the code base without seeking additional approvals or permission or require review. That is the Apache way. Asking the committers very nicely (or write it down) to give a 72h hour review period before merging is perfectly acceptable. Requiring non-Apache tools is not.
> 
> There is nothing wrong with committing directly to gitbox. https://infra.apache.org/apache-github.html claims that it is a dual-master sync setup.

well we can always imagine committers pushing a branch to gitbox if they don't want to use gitbox (tbh we are in 2022 now and not sure this still happen) but why not or providing a patch file attached in a jira (I feel like going back 10 years backward :) ). this has already been mentioned in the discussion here.
But if this happen I don't mind creating a PR from this branch or this patch (as there is a low chance about this happening).
btw the created branch in gitbox will generate a CI build in Jenkins and the branch will be sync to github so from there it's easy to create a PR.
but sure I understand the fact people may not want to create a github account for privacy reason.

**Comment by stephenc** [comment]
> I feel that forcing apache committers to use non-apache (gh) infrastructure by building a process that requires using github might be problematic. If I were on the PMC, I would probably ask the board about this.

I think this is a valid point. If we are documenting a process that uses GH PRs then *at the same time* we should document the process to follow for those who only want to use ASF owned infra (i.e. what do you do if you are using gitbox). This second process will also serve as a fallback if the co-operation with GH ceased (and remember that the ASF is tasked with developing software for the public good, so it has a corporate responsibility to ensure continuity... which is why we have gitbox... even if it can seem like overkill some of the time)

**Comment by michael-o** [comment]
> > I feel that forcing apache committers to use non-apache (gh) infrastructure by building a process that requires using github might be problematic. If I were on the PMC, I would probably ask the board about this.
> 
> I think this is a valid point. If we are documenting a process that uses GH PRs then _at the same time_ we should document the process to follow for those who only want to use ASF owned infra (i.e. what do you do if you are using gitbox). This second process will also serve as a fallback if the co-operation with GH ceased (and remember that the ASF is tasked with developing software for the public good, so it has a corporate responsibility to ensure continuity... which is why we have gitbox... even if it can seem like overkill some of the time)

See: https://github.com/apache/maven-site/pull/321#discussion_r967685008
PR on GH is *one* way, but not the only one. Patch in JIRA issue or branch with review request on dev@ is as good as.

**Comment by olamy** [comment]
> > > I feel that forcing apache committers to use non-apache (gh) infrastructure by building a process that requires using github might be problematic. If I were on the PMC, I would probably ask the board about this.
> > 
> > 
> > I think this is a valid point. If we are documenting a process that uses GH PRs then _at the same time_ we should document the process to follow for those who only want to use ASF owned infra (i.e. what do you do if you are using gitbox). This second process will also serve as a fallback if the co-operation with GH ceased (and remember that the ASF is tasked with developing software for the public good, so it has a corporate responsibility to ensure continuity... which is why we have gitbox... even if it can seem like overkill some of the time)
> 
> See: [#321 (comment)](https://github.com/apache/maven-site/pull/321#discussion_r967685008) PR on GH is _one_ way, but not the only one. Patch in JIRA issue or branch with review request on dev@ is as good as.

correct.
We (Apache) need to be able to accept contributions from everybody without discrimination as some countries may have no access to GitHub see (https://docs.github.com/en/site-policy/other-site-policies/github-and-trade-controls)

**Comment by cstamas** [comment]
Thanks for feedback, but am closing this PR as I have a feeling that this PR will not help, just mess things up, and these are the reasons:
* PR can be created by "anybody" (whoever has a GH account), while doco assumes it is committer only
* in theory I do agree that a code change can be delivered in any way (even with attaching it to JIRA), but only in theory. Personally, I find it very rare to review such a patch attachment, Today a GH pull request -- that I agree, is not "mandatory", but really a "nice thing to have" -- can be reviewed from a handheld, no need for checkout, workspace, IDE, etc.

Moreover, my points were missed IMO, as (in this very moment when I write this comment): 
* the maven-3.8.x branch should be RTC, it exists merely that in (very unlikely, but still) case of some nasty bug report against 3.8.6 force us to provide a patch/fix 3.8.7... IMHO, there should not land any other change,  as it can be easily seen, that it mayeven not produce any release  anymore (or in other words, we do not plan for 3.8.7, but we MAY be forced to make it).
* the maven-3.9.x same thing: we are in few steps before "finish" line, we do plan a release of 3.9.0 in "near future" (was planned for Aug 2022, but heh).... for me, it is RTC only as well
* master is CTR, as release of 4 is frankly, not happening in "near future"

**Comment by michael-o** [comment]
@cstamas I don't think that master should necessarily CTR because there are already a LOT of changes in this branch and chaotic merging and reverts will make it worse. I still consider that even master requires discipline.

**Comment by cstamas** [comment]
And a LOT are yet to land to master, so I don't see why "a LOT of changes present" implies "make it RTC". I don't see why we want to "hold back" the pace of progress (as RTC does exactly that, holds back things, delays them), when our interest would be exactly the opposite: make it happen the soon the better (and this does not imply "chaos" re commits, hopefully).

Also, this is Maven https://github.com/apache/maven we talk about here, but same does not apply to zillion of other Maven repositories.... for example m-ear-p master or whatever...
